File: Data Structures & Algorithms/buy-and-sell-crypto-with-cooldown/submission-1.py
from functools import lru_cache
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        profit = 0
        @lru_cache(None)
        def sol(i,hold):
            # Return the total profit from state (i,hold)
            if i >= len(prices):
                return 0
            if hold == True:
                # sell 
                sell = prices[i] + sol(i+2,False)
                # not sell
                notsell = sol(i+1,hold)

                return max(sell,notsell)

            else:
                # buy 
                buy = -prices[i] + sol(i+1,True)
                notbuy =sol(i+1,hold)

                return max(buy,notbuy)
        return sol(0,0)

File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int postMax = 0;
        int maxProfit = 0;
        for(int i=prices.size()-2;i>=0;i--){
            postMax = max(postMax,prices[i+1]);
            maxProfit = max(postMax - prices[i],maxProfit);

        }
        return maxProfit;

    }
};
